Library/Homebrew/formula.rb

```diff
@@ -1510,7 +1508,20 @@ def eligible_kegs_for_cleanup
           end
         end
       end
-    elsif installed_prefixes.any? && !pinned?
+    elsif installed_prefixes.length > 1
+      if linked_keg.exist?
+        # If the most recent available version is not installed, cleanup
+        # only those older than the currently-linked one.
+        linked_version = PkgVersion.parse(linked_keg.resolved_path.basename.to_s)
+        eligible_for_cleanup = installed_kegs.select { |k| linked_version > k.version }
+        opoo "Skipping (old) #{linked_keg.resolved_path} due to it being linked"
+      else
+        # If no version is linked, keep only the newest one.
```

Comment:
Is this reasonable logic? Should completely unlinked formulas just get cleaned up no matter what? (It would be a LOT more aggressive than the current behavior, though.)